### ISSUE TITLE
PGraphics images bad rendering in retina display using pixelDensity(2)

### ISSUE THREAD

**Issue by gsibaldi** [processing/processing#4530]
Hello, 

it's impossible to use PGraphics with retina display and pixelDensity(2).
If you draw some shape in a PGraphics buffer and then try to display it into the screen using image(pg, x, y) you get a completely wrong result. Images are horizontally stretched by a 2 factor and vertically stretched by a 4 factor with empty horizontal lines interleaved. 

Here is a sample code and a screenshot, I'm on OS X 10.10.2 - Processing 3.1.1

```processing
PGraphics pg;

void setup() 
{
  size(640, 360, JAVA2D);  // not working also in P2D, P3D, FX2D
  pixelDensity(2);
  pg = createGraphics(400, 200); 
}

void draw() 
{ 
  pg.beginDraw();
  pg.background(51);
  pg.noFill();
  pg.stroke(255);
  pg.ellipse(25, 25, 40, 40);
  pg.endDraw();

  image(pg, 120, 60); 
}
```

<img width="639" alt="schermata 2016-06-09 alle 09 40 55" src="https://cloud.githubusercontent.com/assets/14029182/15922142/83c62950-2e26-11e6-8cea-4889e724b368.png">

**Comment by timrolls** [comment]
Having a very similar issue with PGraphics. 
Also running into issues using Capture with pixelDensity. 

Adding pixelDensity(2) to the included background subtraction example looks like:
http://prntscr.com/hmngub

Windows 10
4K screen (3840x2160)

**Comment by benfry** [comment]
@JakubValtar thought you fixed this one when you were working on all the image fixes?

**Comment by JakubValtar** [comment]
@benfry ah, looks like there is a wrong stride, maybe I missed it during testing or it got broken later

I'll look into it

**Comment by JakubValtar** [comment]
Ok, so the original report from @gsibaldi is 3.1.1 which had broken strides on 2x. I believe this is fixed in 3.3.6, if not please reopen this.

@timrolls The problem in your case is that the sketch is doing pixel operations, but they are not adjusted for 2x density. When you call `pixelDensity(2)`, pixel dimensions of the sketch window grow 2x. So your sketch with `size(640, 480)` actually has pixel dimensions `1280x960` (these are stored in `pixelWidth` and `pixelHeight`). Each pixel of the video now needs to cover a square of 2x2 pixels on screen.  Fixed code is here:

```processing
      // ...
      // Render the difference image to the screen
      int c = color(diffR, diffG, diffB);
      if (pixelDensity == 1) {
        // pixels are 1:1
        pixels[i] = c;
      } else {
        // each video pixel needs to cover 2x2 screen pixels
        int x = i % video.width;
        int y = i / video.width;
        pixels[pixelWidth * 2*y + 2*x] = c;
        pixels[pixelWidth * 2*y + (2*x+1)] = c;
        pixels[pixelWidth * (2*y+1) + 2*x] = c;
        pixels[pixelWidth * (2*y+1) + (2*x+1)] = c;
      }
      // ...
```
